### ISSUE TITLE
Use shm-et340 with SMA Energy meter (SHM1.0)

### ISSUE THREAD

**Issue by hoschult** [mitchese/shm-et340#7]
Hi,
tried to use your code with my SMA Energy meter (SHM1.0). Seems to be working - at least partly:

```
DEBU[0019] Received datagram from meter                 
DEBU[0019] Uid: 262816                                  
DEBU[0019] Serial: xxxxxx                           
DEBU[0019] Total W: -7757.2                             
DEBU[0019] Total Buy kWh: 18513.638899999998            
DEBU[0019] Total Sell kWh: 32970.4046                   
INFO[0019] Meter update received: 18513.64 kWh bought and 32970.40 kWh sold, -7757.2 W currently flowing 
DEBU[0019] +-----+-------------+---------------+---------------+ 
DEBU[0019] |value|   L1 	|     L2  	|   L3  	|          
DEBU[0019] +-----+-------------+---------------+---------------+ 
DEBU[0019] |  V  |     1.00 	|     1.00 	|     1.00 	|  
DEBU[0019] |  A  |    -1.70 	|     1.30 	|    -0.70 	|  
DEBU[0019] |  W  |    -1.70 	|     1.30 	|    -0.70 	|  
DEBU[0019] | kWh | 1721346537.27 	| 3285096407.88 	| 4848846278.50 	| 
DEBU[0019] | kWh | 1799534030.79 	| 3363283901.41 	| 4927033772.02 	| 
DEBU[0019] +-----+-------------+---------------+---------------+ 
```

UID, Serial and 'Total W' are correct; V/A/W seems to be not decoded correctly (neither for any phases). Tried to look at the datagrams to be captured of speedwire multicast: https://github.com/datenschuft/SMA-EM/blob/master/speedwiredecoder.py - However I was not able to translate towards your code....

Are you foresee to implement also the readings of SHM1.0 (SMA Energy meter)? That would be fantastic.... I can offer my help by supporting the tests.

Thanks,
Holger

**Comment by hoschult** [comment]
there is a guy at victron community stating the following:

> I have used your GIT hub code to hack my SMA EM v1 to work on this. There is essentially a 4 byte offset on the decodePhaseChunks.

**Comment by hoschult** [comment]
... found a explanation for the 4 byte offset; SHM2 takes also grid frequency into account, SHM1 does not. According to a specification of the speedwire datagram (found at https://openwb.de/forum/viewtopic.php?p=51576 ), grid frequency is at position 160-163:

![image](https://user-images.githubusercontent.com/20515558/188328691-e9a776b6-5051-486e-9bd4-3c6f73b2d753.png)

**Comment by hoschult** [comment]
Got meter working with SMA EM (SHM1); Need to put an offset at decodePhaseChunks in main.go

`        L1 := decodePhaseChunk(b[156:300])
        L2 := decodePhaseChunk(b[300:444])
        L3 := decodePhaseChunk(b[444:588])
`

That's it.... At least for my setup. To do that in a more general way, offset can be determined on the Revision of the EM; Needs to be investigated since which revision frequency (offset) need to be added (or removed).

**Comment by brezel1965** [comment]
I also have an SMA Energy Meter and my consumption is incorrect.  Unfortunately I don't understand how to write this offset in main.go and compile it, maybe someone can provide the modified shm-et340. Thank you Martin

**Comment by hoschult** [comment]
hi Martin. I compiled you the shm10 binary, see download link attached: https://magentacloud.de/s/EdWHd267ffLLkZf

**Comment by buttermaker11111** [comment]
Hello Holger,
can you provide the modified shm-et340 again? Under the link is no file anymore.

Thanks
Olaf

**Comment by brezel1965** [comment]
I get this issue:

2025/06/19 21:28:58 -----------------------------------------------------
2025/06/19 21:28:58 Received datagram from meter at 192.168.0.98:39202
2025/06/19 21:28:58 Uid: 262816
2025/06/19 21:28:58 Serial: 1900234698
2025/06/19 21:28:58 Firmware: 0
2025/06/19 21:28:58 Total W: 3.50
2025/06/19 21:28:58 Total Buy kWh:  57927.5940
2025/06/19 21:28:58 Total Sell kWh: 25885.5639
2025/06/19 21:28:58 +-----+-------------+---------------+---------------+
2025/06/19 21:28:58 |value|   L1 	|     L2  	|   L3  	|
2025/06/19 21:28:58 +-----+-------------+---------------+---------------+
2025/06/19 21:28:58 |  V  |     0.10 	|     0.00 	|     0.00 	|
2025/06/19 21:28:58 |  A  |   -21.87 	|     0.00 	|     0.00 	|
2025/06/19 21:28:58 |  W  |    -2.10 	|     0.70 	|     4.10 	|
2025/06/19 21:28:58 | kWh | 1721346537.26 	| 3285096407.88 	| 4848846278.49 	|
2025/06/19 21:28:58 | kWh | 1799534030.79 	| 3363283901.41 	| 4927033772.02 	|
2025/06/19 21:28:58 +-----+-------------+---------------+---------------+
2025/06/19 21:28:58 Packet (gzip+base64):

**Comment by mitchese** [comment]
ok I guessed wrong, can you give me a sample of the whole packet? It's the part that's not showing.

If you want to see what's in that packet before sending, you can `| base64 -d  | gunzip | hexdump -C` to inspect it...

**Comment by mitchese** [comment]
ah wait, that didn't work with the SMA_ENERGY_METER=true. You should see a line right about the table that it's not SHM2

**Comment by brezel1965** [comment]
oh, I made a mistake when starting, so here again

martin@Linux-Mint:~/Downloads$ SMA_ENERGY_METER=true ./sma_home_manager_printer.linux32 
2025/06/19 21:47:43 Listening for SMA Home Manager packets on 239.12.255.254:9522
2025/06/19 21:49:16 -----------------------------------------------------
2025/06/19 21:49:16 Received datagram from meter at 192.168.0.98:41882
2025/06/19 21:49:16 Uid: 262816
2025/06/19 21:49:16 Serial: .................
2025/06/19 21:49:16 Firmware: 0
2025/06/19 21:49:16 Total W: 3.60
2025/06/19 21:49:16 Total Buy kWh:  57927.5946
2025/06/19 21:49:16 Total Sell kWh: 25885.5639
2025/06/19 21:49:16 SMA Energy Meter, NOT SHM2!
2025/06/19 21:49:16 +-----+-------------+---------------+---------------+
2025/06/19 21:49:16 |value|   L1 	|     L2  	|   L3  	|
2025/06/19 21:49:16 +-----+-------------+---------------+---------------+
2025/06/19 21:49:16 |  V  |  2163.71 	|  3474.43 	|  4785.15 	|
2025/06/19 21:49:16 |  A  |    -3.03 	|    -1.89 	|    -1.37 	|
2025/06/19 21:49:16 |  W  | -6553.59 	| -6553.59 	| -6553.59 	|
2025/06/19 21:49:16 | kWh | 4055746894742.56 	| 726284810909.89 	| 499669251039.60 	|
2025/06/19 21:49:16 | kWh | 218578043972.69 	| 2453305324479.21 	| 3340925475624.04 	|
2025/06/19 21:49:16 +-----+-------------+---------------+---------------+
2025/06/19 21:49:16 Packet (gzip+base64): H4sIAAAJbogA/+zQvU9TYRTH8d9z7xPTexPNNfG2XXptYwf7Mti30bZa21qN0VgHJ+PgoE7VSR1MBxJYSMoO4YYBCBuEoYSFlZa0TEBY6EjDwABhJj0c/gWm811O8ul5cprbfPcC0IYPQBmv4Hz9oR78qtR7wOoISgNAHCowmc9mx0MHBhlgkLkbrf+fYbKZZGp96m8XemLqOzSZtbk3cGCR/YRFlkuMtluw+a1NFrw8/nKK+2Tf4PJ9l36zev7DKIK8HyTz7LOOgxBbiAxvHlW6CPP9MJl+3z4KIML3I2T2+UnsAzx+65F5n15u+XhM5iOqAXO6iRj/nwTvJmjXijfMEZJsSTLj31yhjRRbigxubzGANFuaTC8d7HaRYcuQ2YPh2x1k2bJk5r0nsX3k2PI8Czyf307aTYavVmZQZCuSqeXx2h+U2Epkphrk2yizlcmM3wv9MqpsVbJU/XD+KWpsNTLVdy6iqLO95tm4mZ3J1zX0R0iSJEmSJEmSJEmSJEmSJEmSJEmSJEl33vUAtHUjgQAgAAA=
2025/06/19 21:49:17 -----------------------------------------------------

**Comment by brezel1965** [comment]
Yes, you've identified it correctly. Currently, there's only one phase, L1, connected to the 230V network. (Grüsse nach Köln ;-) ) Phases L2 and L3 are currently disconnected.

The total purchase and total sale values ​​are correct, as are the values ​​for L1, L2, and L3; I get the same values ​​in FHEM.
